### ISSUE TITLE
Cannot read property 'indexOf' of undefined at Function.isPrematureTemplateContentError

### ISSUE THREAD

**Issue by reubenlillie** [11ty/eleventy#1654]
**Describe the bug**
I cloned a fresh copy of my portfolio site repo, and now Eleventy is throwing an `isPrematureTemplateContentError`, whether building or serving.

**To Reproduce**

1. Clone my portfolio site repository

```cli
git clone --recurse-submodules https://gitlab.com/reubenlillie/reubenlillie.com.git
```

2. Install Eleventy and dependencies

```cli
npm install
```

3. Run Eleventy

```cli
npm run dev
```

or 

```cli
npm run build
```

4. See error:

```cli
Problem writing Eleventy templates: (more in DEBUG output)
> Cannot read property 'indexOf' of undefined

`TypeError` was thrown:
    TypeError: Cannot read property 'indexOf' of undefined
        at Function.isPrematureTemplateContentError (~/reubenlillie.com/node_modules/@11ty/eleventy/src/EleventyErrorUtil.js:69:17)
        at Template.render (~/reubenlillie.com/node_modules/@11ty/eleventy/src/TemplateContent.js:207:29)
        at async Template._getContent (~/reubenlillie.com/node_modules/@11ty/eleventy/src/Template.js:547:12)
        at async Template.getTemplateMapContent (~/reubenlillie.com/node_modules/@11ty/eleventy/src/Template.js:732:19)
        at async TemplateMap.populateContentDataInMap (~/reubenlillie.com/node_modules/@11ty/eleventy/src/TemplateMap.js:359:39)
        at async TemplateMap.cache (~/reubenlillie.com/node_modules/@11ty/eleventy/src/TemplateMap.js:303:5)
        at async TemplateWriter._createTemplateMap (~/reubenlillie.com/node_modules/@11ty/eleventy/src/TemplateWriter.js:170:5)
        at async TemplateWriter.writeTemplates (~/reubenlillie.com/node_modules/@11ty/eleventy/src/TemplateWriter.js:203:5)
        at async TemplateWriter.write (~/reubenlillie.com/node_modules/@11ty/eleventy/src/TemplateWriter.js:254:25)
        at async Eleventy.write (~/reubenlillie.com/node_modules/@11ty/eleventy/src/Eleventy.js:743:13)
Copied 50 files / Wrote 0 files in 0.68 seconds (v0.11.1)
```

**Expected behavior**
A build

**Environment:**

- OS and Version: Ubuntu 20.04.2 LTS 64-bit
- Eleventy Version 0.11.1

**Additional context**
See latest Netlify build output: https://app.netlify.com/sites/reubenlillie/deploys/6032f9782e6d290008e69d73

**Comment by pdehaan** [comment]
Curious curious...

I was able to reproduce on 0.10.0, 0.11.0, and 0.11.1 (didn't try v1.0.0-canary -- v0.10 might have been a different error).
I tried removing ESM plumbing but got same errors.
I tried removing 99% of the pages and eventually it worked (rendering only the 404 template).

Looks like the error has something to do w/ a combination of cloudinary and eleventy-img. Even if I update the eleventy-img plugin to the latest version, I still had problems. But if I short circuit all the image plugin stuff and just return `<img src="HEADSHOT.png"/>` early, it eventually seems to work (if I ran <kbd>npm run dev:11ty</kbd> directly, to avoid ESLint warnings errors I didn't want to bother debugging).
Not sure if there was some breaking change in that plugin that broke you downstream, or I'm just seeing bogus errors locally because I don't have cloudinary API keys/tokens and that's breaking something else.

But that's where I'd start. Hack out the two references to eleventy-img and see if that builds locally for you. Probably easiest to rename the original files, and put in temporary shortcode files that just render a static string:

```sh
git grep -n " from '@11ty/eleventy-img'"

_includes/shortcodes/background-image.js:14: import Image from '@11ty/eleventy-img'
_includes/shortcodes/headshot.js:10: import Image from '@11ty/eleventy-img'
```

**Comment by pdehaan** [comment]
For the record, [Eleventy debugging](https://www.11ty.dev/docs/debugging/) wasn't really any help whatsoever.

If you do get it working without Cloudinary and [eleventy-img](https://www.11ty.dev/docs/plugins/image/) code, it'd be interesting to add those back and add some bonus `try..catch` blocks around anything async and see if that gives you any more relevant debugging information.

**Comment by reubenlillie** [comment]
Ah, I forgot to add a .env file. I'll test soon to confirm. But yes, the error threw me. Thanks for shaking that memory loose.

**Comment by reubenlillie** [comment]
Thanks, @pdehaan. It was, in fact, an environment variable issue. I added a note to my README to remind of this moment in the future. I’m not sure if this is an opportunity to consider better error/debugging messages. I’m grateful for your help all the same.

**Comment by pdehaan** [comment]
I haven't had too much time to dig into it, but I was assuming this was an eleventy-img bug with insufficient logging on a possibly undefined `src` attribute so I could file it upstream. But early tests show that you in fact get excellent debug information if you use the README examples of `{% image src %}` shortcodes without a defined source. I'd have to dig in further to see if its possibly an issue w/ LiquidJS templates vs .11ty.js templates w/ exported `data` and `render` props like you were using, or how something else was interacting with your headshot and background images.

If I get more time/motivation I can try recloning your repo and trying some extra try/catch blocks or something else to get more useful error logging. Not sure if it should fall to Eleventy to intercept errors and log them vs swallowing them, or if people should be wrapping their own async code in try/catch blocks like you would outside of Eleventy.

**Comment by pdehaan** [comment]
@reubenlillie I know you already solved this puzzle, but turns out the problem was with **cloudinary** API not liking the missing env vars (not **eleventy-img**, like I originally suspected). I wrapped a `try..catch` around each of the async functions in headshot.js and background-image.js and it looks like this was culprit:

```js
  var src;
  try {
    src = cloudinary.v2.url(source, {
      secure: true, 
      transformation: [
        {
          quality: 'auto',
        }
      ]
    })
  } catch (err) {
    console.error(err);
    return;
  }
```

Now I can "successfully" build the site w/ the following output:

```sh
Unknown cloud_name
Unknown cloud_name

Writing _site/index.html from ./index.11ty.js.
Writing _site/404.html from ./404.11ty.js.
Writing _site/contact/index.html from ./content/pages/contact.11ty.js.
Writing _site/offline/index.html from ./content/pages/offline.11ty.js.
Writing _site/thanks/index.html from ./content/pages/thanks.11ty.js.
Writing _site/accessibility/index.html from ./content/pages/policies/accessibility.11ty.js.
Writing _site/colophon/index.html from ./content/pages/policies/colophon.11ty.js.
Writing _site/copyright/index.html from ./content/pages/policies/copyright.11ty.js.
Writing _site/privacy/index.html from ./content/pages/policies/privacy.11ty.js.
Writing _site/terms/index.html from ./content/pages/policies/terms.11ty.js.
Writing _site/concert/index.html from ./content/pages/singing/concert.11ty.js.
Writing _site/stage/index.html from ./content/pages/singing/stage.11ty.js.
Writing _site/singing/index.html from ./content/pages/singing/index.11ty.js.
Writing _site/speaking/index.html from ./content/pages/speaking/index.11ty.js.
Writing _site/articles/index.html from ./content/pages/writing/articles.11ty.js.
Writing _site/curriculum-vitae/index.html from ./content/pages/writing/curriculum-vitae.11ty.js.
Writing _site/papers/index.html from ./content/pages/writing/papers.11ty.js.
Writing _site/software/index.html from ./content/pages/writing/software.11ty.js.
Writing _site/baritone-rep/index.html from ./content/pages/singing/baritone-rep.md.
Writing _site/teaching/index.html from ./content/pages/teaching/index.md.
Writing _site/parliamentary-procedure/index.html from ./content/pages/teaching/parliamentary-procedure.md.
Writing _site/writing/index.html from ./content/pages/writing/index.md.
Writing _site/syllabi/index.html from ./content/pages/teaching/syllabi/index.11ty.js.
Writing _site/comp-496/index.html from ./content/pages/teaching/syllabi/comp-496.md.

Copied 50 files / Wrote 24 files in 0.42 seconds (17.5ms each, v0.11.1)
```

TL;DR: I guess wrapping any async calls or any code that can throw unexpected errors w/ `try..catch` will help save future debugging issues since a  better error message would have helped us track this down much quicker.

❤️

**Comment by reubenlillie** [comment]
Thanks, @pdehaan. That's helpful. I agree. I've got `try . . . catch` statements elsewhere. But I missed this one, and I re-learned that lesson as a result.
